File: src/data/notebook.test.ts
import assert from 'node:assert/strict'
import test from 'node:test'
import {
  editorJsonToPlainText,
  ensureNotebookEditor,
  normaliseNotebookEditorJson,
  plainTextToEditorJson,
  type Note,
  type NotebookEditorJson,
} from './notebook.ts'

test('legacy markdown is preserved but its formatting marks do not enter the readable editor', () => {
  const note: Note = {
    id: 'legacy', title: 'Legacy', body: '# Heading\n- Point\n> Quote', tags: [], updatedAt: '2026-01-01T00:00:00.000Z',
  }
  const migrated = ensureNotebookEditor(note)
  assert.equal(migrated.legacyMarkdownSource, note.body)
  assert.equal(editorJsonToPlainText(migrated.editorJson), 'Heading\nPoint\nQuote')
})

test('new text produces a Lexical-compatible serialised root', () => {
  const state = plainTextToEditorJson('First\nSecond')
  assert.equal(state.root.type, 'root')
  assert.equal(state.root.version, 1)
  assert.equal(state.root.children[0].type, 'paragraph')
  assert.equal(state.root.children[0].children?.[0].type, 'text')
  assert.equal(editorJsonToPlainText(state), 'First\nSecond')
})

test('the temporary block JSON migrates lazily without losing its words', () => {
  const old = {
    root: {
      type: 'root', version: 1,
      children: [
        { type: 'heading', version: 1, text: 'Clinical sequence' },
        { type: 'paragraph', version: 1, text: 'Assess airway first.' },
      ],
    },
  } as NotebookEditorJson
  const next = normaliseNotebookEditorJson(old)
  assert.equal(editorJsonToPlainText(next), 'Clinical sequence\nAssess airway first.')
  assert.ok(next.root.children.every((node) => Array.isArray(node.children)))
})

File: android/app/src/test/java/com/synapse/app/core/notebook/NotebookTest.kt
package com.synapse.app.core.notebook

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * Test vectors ported verbatim from `src/data/notebook.test.ts` so Android's
 * pure notebook transforms stay numerically/structurally identical to web's.
 */
class NotebookTest {

    private val json = Json { ignoreUnknownKeys = true }

    // "legacy markdown is preserved but its formatting marks do not enter the readable editor"
    @Test
    fun legacyMarkdownIsPreservedButItsFormattingMarksDoNotEnterTheReadableEditor() {
        val note = Note(
            id = "legacy",
            title = "Legacy",
            body = "# Heading\n- Point\n> Quote",
            updatedAt = "2026-01-01T00:00:00.000Z",
        )

        val migrated = ensureNotebookEditor(note)

        assertEquals(note.body, migrated.legacyMarkdownSource)
        assertEquals("Heading\nPoint\nQuote", editorJsonToPlainText(migrated.editorJson))
    }

    // "new text produces a Lexical-compatible serialised root"
    @Test
    fun newTextProducesALexicalCompatibleSerialisedRoot() {
        val state = plainTextToEditorJson("First\nSecond")

        assertEquals("root", state["root"]!!.jsonObject["type"]!!.jsonPrimitive.content)
        assertEquals(1, state["root"]!!.jsonObject["version"]!!.jsonPrimitive.content.toInt())
        val firstChild = state["root"]!!.jsonObject["children"]!!.jsonArray[0].jsonObject
        assertEquals("paragraph", firstChild["type"]!!.jsonPrimitive.content)
        assertEquals("text", firstChild["children"]!!.jsonArray[0].jsonObject["type"]!!.jsonPrimitive.content)
        assertEquals("First\nSecond", editorJsonToPlainText(state))
    }

    // "the temporary block JSON migrates lazily without losing its words"
    @Test
    fun theTemporaryBlockJsonMigratesLazilyWithoutLosingItsWords() {
        val old = json.parseToJsonElement(
            """
            {
              "root": {
                "type": "root", "version": 1,
                "children": [
                  { "type": "heading", "version": 1, "text": "Clinical sequence" },
                  { "type": "paragraph", "version": 1, "text": "Assess airway first." }
                ]
              }
            }
            """.trimIndent(),
        ).jsonObject

        val next = normaliseNotebookEditorJson(old)

        assertEquals("Clinical sequence\nAssess airway first.", editorJsonToPlainText(next))
        val children = next["root"]!!.jsonObject["children"]!!.jsonArray
        assertTrue(children.all { (it as JsonObject)["children"] != null })
    }
}
